Gutenboarding: Add auth store for managing login flow state

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

pbAok1-od-p2

This PR adds a new "auth" store to the `@automattic/data-stores` package. This store makes the necessary API calls to log a user in, and manages the state of where the user is up to in the login flow.

High level it's modelling a finite state machine. E.g. the user submits their username and the store will transition to the next state depending on what auth method the user is using.

See the README for an example of what this might look like from the component's point of view.

Please let me know what you think of the design. My motivation was keeping more of the business logic in the store rather than spread throughout the lifecycle methods of the UI (which is what seemed to happen in the old signup framework I think, despite the efforts of the original architects).

My plan with this PR isn't to get every flow fully implemented before merging. I want to get feedback, decide whether this is the right direction, and if so add more flows in future PRs.

Also:
- Added `qs` to `package.json`, which was actually already being used e.g. `client/state/login/utils.js`
- Added `env=browser` to eslint config to use `fetch()` without warnings

#### Testing instructions

##### Auto

```
npm run test-packages -- --testPathPattern packages/data-store
```

##### Manual

Added some console commands to make testing a little easier

* Navigate to `/gutenboarding`
* Open network and console in dev tools
* Trigger any store action with `wp.auth.actionName( 'actionParams' )` e.g.
  * `wp.auth.submitUsernameOrEmail( 'user1' )`
  * `wp.auth.submitPassword( 'passw0rd' )`
  * `wp.auth.reset()`
* changes to `loginFlowState` will be printed to the console
* Use network tab inspect what requests/responses look like

p-jackson committed:
Add auth store for managing login flow state

andrewserong commented on packages/data-stores/src/auth/actions.ts:
```diff
@@ -0,0 +1,87 @@
+	}
+}
+
+export const recieveWpLogin = ( response: WpLoginSuccessResponse ) =>
```
Typo in this and `recieveWpLoginFailed` (should be `receive`)

andrewserong commented on packages/data-stores/src/auth/reducer.ts:
```diff
@@ -0,0 +1,63 @@
+	}
+};
+
+const password: Reducer< string, Action > = ( state = '', action ) => {
```
Do we need a password reducer? Looks like we're not (yet) storing anything here, and it seems like the `submitPassword` and `wpLogin` actions work nicely without it? (I always feel wary of storing passwords in state if we don't need to)

p-jackson replied:
Ya! I don't need this anymore, and I agree I don't want to naïvely JSON dump the store or something and leak it without realising.

andrewserong reviewed (commented):
Nice work on this @p-jackson, I'm really liking the design, and you make an excellent case for the separate data store with the complexity of managing the business logic. I think it makes great sense to keep the business logic in the data store so that the consuming code can be more concerned with the UI — it feels like a good way to ensure the stability of the login behaviour. It also feels cleaner to me that this auth store doesn't include account creation. Great addition with the unit tests, too!

I like the pattern of using the `getLoginFlowState` and matching against string constants for which part of the flow a user is in. It's a good use case for TS too, here, to make it easier for folks writing consuming code.

I didn't have time to do a close review of this, but added a couple of tiny comments, but otherwise just wanted to say the direction feels good to me! 👍

roo2 commented on packages/data-stores/src/auth/controls.ts:
```diff
@@ -0,0 +1,39 @@
+		},
+		WP_LOGIN: async ( { action, params }: ReturnType< typeof wpLogin > ) => {
+			return await wpcomRequest( {
+				path: '/wp-login.php?action=' + action,
```
Since we only call `WP_LOGIN` from the one place, I would consider in-lining the action so it hits `/wp-login.php?action=login-endpoint` every time.

p-jackson replied:
I was thinking of re-using the `WP_LOGIN` control for the `wp-login.php?action=two-step-authentication-endpoint` request too, which is why I didn't hard code the query param.

But maybe re-using the same control for both cases will be too messy 🤔 I might make seperate `LOGIN_ENDPOINT` and `TWO_STEP_AUTHENTICATION_ENDPOINT` controls but I'm not sure yet.

roo2 replied:
yea up to you, I think passing the param into a URL always makes me a bit uncomfortable because of things like `?action=` +`action&param2=something`. It's harmless if that does occur because `wp-login.php` is a public endpoint anyway... and we would be hardcoding `action` to an action name string anyway... it just makes me feel slight discomfort for some reason :joy:

p-jackson replied:
Excellent point! Should have thought more of the evils of string concat!

I'll see if there's a real url builder library already in the `package.json`, but a seperate control might just be simpler.

roo2 replied:
Maybe `encodeURIComponent()` is the best way to go... or if you can have a type on action so it can only be `two-step-authentication-endpoint | login-endpoint`. Yea but it's not necessary for code/security just for OCD satisfaction.

p-jackson replied:
Now I've seen it I can't unsee it lol

roo2 commented on packages/data-stores/src/auth/README.md:
```diff
@@ -0,0 +1,83 @@
+- `ENTER_2FA_CODE`
+- `LOGIN_LINK_SENT`
+- `WAITING_FOR_2FA_APP`
+- `LOGGED_IN`
```
hmm I've got a few questions about the states: 

Are all of the states definitely mutually exclusive? e.g. Can a user be in ENTER_2FA_CODE and LOGIN_LINK_SENT at the same time? 

I noticed that you can access the existing `log-in` page while already being `LOGGED_IN` in to an existing account and it gives you the option to e.g. "Continue as Eru Penkman" or log in as someone different.

Is the plan to have the states persisted between browser reloads? e.g. if a user were in the ENTER_2FA_CODE state and reloaded the app, what should happen? What if they're in the LOGGED_IN state?

p-jackson replied:
> Are all of the states definitely mutually exclusive? e.g. Can a user be in ENTER_2FA_CODE and LOGIN_LINK_SENT at the same time? 

Hmm these really are more UI states then they are ... states of the world (for want of a better word!) The `LOGIN_LINK_SENT` state is supposed mean "you should show the login link sent UI", not "in the past a login link has been sent".

Maybe part of the problem is what I called it: `loginFlowState`. I've been trying to think of something better 😅

> I noticed that you can access the existing `log-in` page while already being `LOGGED_IN` in to an existing account and it gives you the option to e.g. "Continue as Eru Penkman" or log in as someone different.

Nice spotting! Hadn't thought of that case.

What do you think of the auth store ignoring existing sessions? The login components could use the user store to show the the "Continue as Alex" button?

> Is the plan to have the states persisted between browser reloads?

The current `/log-in` UI throws all the state away when the browser reloads. My first reaction is just to copy that behaviour. For gutenboarding it'll be in the modal, and the modal will also close when the page refreshes.

But if we wanted to change that I think it's just a flag you can pass to `registerStore()`

roo2 replied:
ahh nice both of those answers make sense, yea so it's more the state of the login page then the state of the world.. yea I do think `loginFlowState` is a good name.

And yep that sounds like it would work to have the auth store ignore any existing session and managing the already logged in case from the component level.

👍

ramonjd replied:
❤️ the idea of using states. I wonder if we could use something similar for https://github.com/Automattic/wp-calypso/issues/39268

For example, if a user is in the `NEW_SITE_CREATED` state we know not to allow site creation.

Or something.

roo2 reviewed (commented):
This looks awesome 👍 I had some questions about how the state machine would work in some edge cases

ramonjd commented on packages/data-stores/src/user/README.md:
```diff
@@ -2,6 +2,8 @@
 
 This store includes behavior for checking if a current user is logged in, and creating new user accounts.
 
+To log the user in see the [auth store](../auth/README.md).
```
You thought of everything 🙇

p-jackson committed:
Fix typos

p-jackson committed:
Remove password from store

We don't need to store it yet, so lets not keep sensitive info stored in
memory when we don't need to

p-jackson committed:
Escape the user's email in API requests

p-jackson committed:
Fix typo in auth-options path

p-jackson committed:
Add dev tools for testing auth store

p-jackson committed:
Using fetch() to call wp-login.php

This copies how it's done in
`client/state/login/utils.js`

Added "qs" to package.json, which was already being used in
`client/state/login/utils.js`

Choose 6.9.1 because that's the version being used in
`client/package.json`

p-jackson committed:
Mock fetch in tests

p-jackson committed:
Update comment

p-jackson committed:
Let eslint know fetch is defined

p-jackson committed:
Protect against arbitrary strings being appended to wp-login url

p-jackson commented:
Ready for review. Added some testing instructions.

andrewserong commented on packages/data-stores/src/auth/actions.ts:
```diff
@@ -0,0 +1,100 @@
+	const username = yield { type: 'SELECT_USERNAME_OR_EMAIL' };
+
+	try {
+		const loginResponse = yield fetchWpLogin( 'login-endpoint', { username, password } );
```
This returns a JSON blob even if a login fails (e.g. incorrect password with code `incorrect_password`), so the below `receiveWpLogin` call means that when I run `wp.auth.submitPassword( 'incorrect-passw0rd' );` in the console, I get `New loginFlowState = LOGGED_IN`, where it should error out.

p-jackson replied:
Hmmm, yeah was obviously expecting a throw here, but you're right my intention was to do error cases et al. in subsequent PRs.

Cheers all!

andrewserong reviewed (commented):
This is looking good @p-jackson — I really like the direction this data store is going in. I started to review the submit password behaviour (I've left a single comment) and then it occurred to me that your plan is probably to add in error handling, etc, in subsequent PRs and just get the initial body of work merged in to make the further changes easier to push up, review and merge in as you go. If so, this looks good to merge in to me!

sirreal commented on packages/data-stores/src/auth/controls.ts:
```diff
@@ -0,0 +1,49 @@
+				path: `/users/${ escaped }/auth-options`,
+				apiVersion: '1.1',
+			} );
+		},
```
This pattern proliferating across stores, I'd really love to see the generic control for `wpcom-proxy-request` move forward:

https://github.com/Automattic/wp-calypso/pull/38959

p-jackson replied:
Sounds good, would the idea be (if a store needs multiple controls) to spread the generic control into this stores? e.g.

```typescript
import { controls as genericControls } from '../wpcom-request-controls';

export const controls = {
	...genericControls,
	MY_STORE_CONTROL: () => {}
};
```

Or is there another way to share controls across stores?

sirreal replied:
Yes, I believe we'd merge controls. I believe we'll find that all the controls are generic for handling a small set of side-effects, and the specific functionality will move into action creators. This should also make testing much easier and reduce the need for mocking in many places.

p-jackson replied:
Sounds good, I do like less mocking :)

sirreal commented on packages/data-stores/src/auth/controls.ts:
```diff
@@ -0,0 +1,49 @@
+
+			return await response.json();
+		},
+	};
```
A shared, plain `fetch` control that sets typical defaults like you've done here would also be nice 🙂 Most of our usage of `apiFetch` should either be wpcom-proxy-request or a plain fetch, I believe continued usage of the `apiFetch` singleton will only become more complicated and I'd prefer to migrate away from it.
